internal/engine/experiment/tlsmiddlebox/tracing.go

```diff
@@ -0,0 +1,142 @@
+package tlsmiddlebox
+
+//
+// Iterative network tracing
+//
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"net"
+	"sort"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/ooni/probe-cli/v3/internal/measurexlite"
+	"github.com/ooni/probe-cli/v3/internal/model"
+	"github.com/ooni/probe-cli/v3/internal/netxlite"
+	utls "gitlab.com/yawning/utls.git"
+)
+
+// ClientIDs to map configurable inputs to uTLS fingerprints
+// We use a non-zero index to map to each ClientID
+var ClientIDs = map[int]*utls.ClientHelloID{
+	1: &utls.HelloGolang,
+	2: &utls.HelloChrome_Auto,
+	3: &utls.HelloFirefox_Auto,
+	4: &utls.HelloIOS_Auto,
+}
+
+// MeasureTLS performs tracing using control and target SNI
+func (m *Measurer) TLSTrace(ctx context.Context, index int64, zeroTime time.Time, logger model.Logger,
+	address string, targetSNI string, trace *CompleteTrace) {
+	// perform an iterative trace with the control SNI
+	trace.ControlTrace = m.IterativeTrace(ctx, index, zeroTime, logger, address, m.config.snicontrol())
+	// perform an iterative trace with the target SNI
+	trace.TargetTrace = m.IterativeTrace(ctx, index, zeroTime, logger, address, targetSNI)
+}
+
+// IterativeTrace creates a Trace and calls iterativeTrace
+func (m *Measurer) IterativeTrace(ctx context.Context, index int64, zeroTime time.Time, logger model.Logger,
+	address string, sni string) (tr *IterativeTrace) {
+	tr = &IterativeTrace{
+		SNI:        sni,
+		Iterations: []*Iteration{},
+	}
+	maxTTL := m.config.maxttl()
+	m.iterativeTrace(ctx, index, zeroTime, logger, address, sni, maxTTL, tr)
+	tr.Iterations = alignIterations(tr.Iterations)
+	return
+}
+
+// iterativeTrace performs iterative tracing with increasing TTL values
+func (m *Measurer) iterativeTrace(ctx context.Context, index int64, zeroTime time.Time, logger model.Logger,
+	address string, sni string, maxTTL int64, trace *IterativeTrace) {
+	ticker := time.NewTicker(m.config.delay())
+	wg := new(sync.WaitGroup)
+	for i := int64(1); i <= maxTTL; i++ {
+		wg.Add(1)
+		go m.handshakeWithTTL(ctx, index, zeroTime, logger, address, sni, int(i), trace, wg)
+		<-ticker.C
+	}
+	wg.Wait()
+}
+
+// HandshakeWithTTL performs the TLS Handshake using the passed ttl value
+func (m *Measurer) handshakeWithTTL(ctx context.Context, index int64, zeroTime time.Time, logger model.Logger,
+	address string, sni string, ttl int, tr *IterativeTrace, wg *sync.WaitGroup) {
```

Comment:
Because this algorithm is a bit long a suggest to use vertical spaces to split relevant parts and add a comment before each part concisely saying what this part is doing. It will improve readability.